NineChronicles.Headless/Middleware/CustomRateLimitMiddleware.cs

```diff
@@ -0,0 +1,61 @@
+using System.IO;
+using System.Threading.Tasks;
+using AspNetCoreRateLimit;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using Serilog;
+using ILogger = Serilog.ILogger;
+
+namespace NineChronicles.Headless.Middleware
+{
+    public class CustomRateLimitMiddleware : RateLimitMiddleware<CustomIpRateLimitProcessor>
+    {
+        private readonly ILogger _logger;
+        private readonly IRateLimitConfiguration _config;
+
+        public CustomRateLimitMiddleware(RequestDelegate next,
+            IProcessingStrategy processingStrategy,
+            IOptions<IpRateLimitOptions> options,
+            IIpPolicyStore policyStore,
+            IRateLimitConfiguration config)
+            : base(next, options?.Value, new CustomIpRateLimitProcessor(options?.Value!, policyStore, processingStrategy), config)
+        {
+            _config = config;
+            _logger = Log.Logger.ForContext<CustomRateLimitMiddleware>();
+        }
+
+        protected override void LogBlockedRequest(HttpContext httpContext, ClientRequestIdentity identity, RateLimitCounter counter, RateLimitRule rule)
+        {
+            _logger.Information($"[IP-RATE-LIMITER] Request {identity.HttpVerb}:{identity.Path} from IP {identity.ClientIp} has been blocked, " +
+                                $"quota {rule.Limit}/{rule.Period} exceeded by {counter.Count - rule.Limit}. Blocked by rule {rule.Endpoint}, " +
+                                $"TraceIdentifier {httpContext.TraceIdentifier}. MonitorMode: {rule.MonitorMode}");
+        }
+
+        public override async Task<ClientRequestIdentity> ResolveIdentityAsync(HttpContext httpContext)
+        {
+            var identity = await base.ResolveIdentityAsync(httpContext);
+            identity.ClientIp = identity.ClientIp + "/" + httpContext.Connection.RemotePort;
+
+            if (httpContext.Request.Protocol == "HTTP/2")
+            {
+                return identity;
+            }
+
+            if (httpContext.Request.Protocol == "HTTP/1.1")
+            {
+                httpContext.Request.EnableBuffering();
+                var body = await new StreamReader(httpContext.Request.Body).ReadToEndAsync();
+                httpContext.Request.Body.Seek(0, SeekOrigin.Begin);
+                if (body.Contains("stageTransaction"))
+                {
+                    identity.Path = "/graphql/stagetransaction";
+                }
+
+                return identity;
+            }
+
+            identity.ClientIp = "127.0.0.1";
```

Comment:
Is there a reason to update the client IP to localhost manually?

Reply:
Oh, this was to manually "whitelist" other http protocols so that they wouldn't be rate limited. But, I think they won't be affected by these rules(https://github.com/planetarium/NineChronicles.Headless/pull/1999/files#diff-e47fc585cf3ec48a24eabba8519cf4fcffbfc15f80cc907dec2379af8ebdf134R102-R120) anyway 😅. I'll take this one out. Thanks for checking!